src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -725,27 +830,56 @@ private ValueTask<FlushResult> TimeFlushUnsynchronizedAsync()
         return _flusher.FlushAsync(_minResponseDataRate, bytesWritten);
     }
 
-    public bool TryUpdateConnectionWindow(int bytes)
+    private (long, long) ConsumeWindow(long bytes)
     {
-        lock (_writeLock)
+        lock (_windowUpdateLock)
         {
-            return _connectionOutputFlowControl.TryUpdateWindow(bytes);
+            var actual = Math.Min(bytes, _window);
+            var remaining = _window -= actual;
+
+            return (actual, remaining);
         }
     }
 
-    public bool TryUpdateStreamWindow(StreamOutputFlowControl flowControl, int bytes)
+    private void AbortFlowControl()
     {
-        lock (_writeLock)
+        lock (_windowUpdateLock)
         {
-            return flowControl.TryUpdateWindow(bytes);
+            while (_waitingForMoreWindow.TryDequeue(out var producer))
+            {
+                if (!producer.StreamCompleted)
+                {
+                    // Stop the output
+                    producer.Stop();
+                }
+            }
         }
     }
 
-    public void AbortPendingStreamDataWrites(StreamOutputFlowControl flowControl)
+    public bool TryUpdateConnectionWindow(int bytes)
     {
-        lock (_writeLock)
+        lock (_windowUpdateLock)
         {
-            flowControl.Abort();
+            var maxUpdate = Http2PeerSettings.MaxWindowSize - _window;
+
+            if (bytes > maxUpdate)
+            {
+                return false;
+            }
+
+            _window += bytes;
+
+            while (_waitingForMoreWindow.TryDequeue(out var producer))
+            {
+                if (!producer.StreamCompleted)
+                {
+                    // We're no longer waiting for the update
+                    producer.MarkWaitingForWindowUpdates(false);
```

Comment:
OK I spent some time with this today looking at the most minimal changes to basically replicate what OutputFlowControl did. Let me know if my assumptions are wrong in any of this.

The old code also scheduled all waiting streams as long as while there was still flow control window available. I don't see any special logic to deal with fairness outside of that. 

https://github.com/dotnet/aspnetcore/blob/1d922b9dbdaeafe7468094568f934cbcc8c8dc73/src/Servers/Kestrel/Core/src/Internal/Http2/FlowControl/OutputFlowControl.cs#L49-L62

So when CompleteCurrent is called, it queues the continuation to the threadpool and keeps doing so while there's flow window available. I'm missing how this avoids the thundering herd problem.

Reply:
Actually, looking at it more closely, since TryUpdateWindow for the connection level flow control is called within the write lock, I'm not sure how the available window can be decremented (via call to OutputFlowControl.Advance) since its also only allowed to be called inside of the write lock. It seems like all streams would need to be dequeued before any new stream can proceed since that happens while the lock is held.

I'm not sure `_flow.Available > 0` is ever false.

Reply:
I added a commit to do something minimal that I think solves the problem in the worst case. Please review and let me know if I missed something about the existing code.